packages/perseus-core/src/utils/util.ts

```diff
@@ -0,0 +1,688 @@
+    const span = range[1] - range[0];
+    const numTicks = span / step;
+    if (numTicks <= 10) {
+        // Will displays fine on mobile
+        return step;
+    }
+    if (numTicks <= 20) {
+        // Will be crowded on mobile, so hide every other tick
+        return step * 2;
+    }
+    // Fallback in case we somehow have more than 20 ticks
+    // Note: This shouldn't happen due to GraphSettings.validStep
+    return tickStepFromNumTicks(span, 10);
+};
+
+/**
+ * Constrain tick steps intended for desktop size graphs
+ * to something more suitable for mobile size graphs.
+ * Specifically, we aim for 10 or fewer ticks per graph axis.
+ */
+function constrainedTickStepsFromTickSteps(
+    tickSteps: [number, number],
+    ranges: [Range, Range],
+): Coordinates {
+    return [
+        constrainTickStep(tickSteps[0], ranges[0]),
+        constrainTickStep(tickSteps[1], ranges[1]),
+    ];
+}
+
+/**
+ * Approximate equality on numbers and primitives.
+ */
+function eq<T>(x: T, y: T): boolean {
+    if (typeof x === "number" && typeof y === "number") {
+        return Math.abs(x - y) < 1e-9;
+    }
+    return x === y;
+}
+
+/**
+ * Deep approximate equality on primitives, numbers, arrays, and objects.
+ * Recursive.
+ */
+function deepEq<T>(x: T, y: T): boolean {
+    if (Array.isArray(x) && Array.isArray(y)) {
+        if (x.length !== y.length) {
+            return false;
+        }
+        for (let i = 0; i < x.length; i++) {
+            if (!deepEq(x[i], y[i])) {
+                return false;
+            }
+        }
+        return true;
+    }
+    if (Array.isArray(x) || Array.isArray(y)) {
+        return false;
+    }
+    if (typeof x === "function" && typeof y === "function") {
+        return eq(x, y);
+    }
+    if (typeof x === "function" || typeof y === "function") {
+        return false;
+    }
+    if (typeof x === "object" && typeof y === "object" && !!x && !!y) {
+        return (
+            x === y ||
+            (_.all(x, function (v, k) {
+                // @ts-expect-error - TS2536 - Type 'CollectionKey<T>' cannot be used to index type 'T'.
+                return deepEq(y[k], v);
+            }) &&
+                _.all(y, function (v, k) {
+                    // @ts-expect-error - TS2536 - Type 'CollectionKey<T>' cannot be used to index type 'T'.
+                    return deepEq(x[k], v);
+                }))
+        );
+    }
+    if ((typeof x === "object" && !!x) || (typeof y === "object" && !!y)) {
+        return false;
+    }
+    return eq(x, y);
+}
+
+/**
+ * Query String Parser
+ *
+ * Original from:
+ * http://stackoverflow.com/questions/901115/get-querystring-values-in-javascript/2880929#2880929
+ */
+function parseQueryString(query: string): QueryParams {
+    // TODO(jangmi, CP-3340): Use withLocation to access SSR safe location.
+    // eslint-disable-next-line no-restricted-syntax
+    query = query || window.location.search.substring(1);
```

Comment:
I think this is safe to move, but we should probably drop the `|| window...` bit as that won't be safe in a non-browser env.